retrieval/discovery/kubernetes/discovery.go

```diff
@@ -215,11 +229,78 @@ func (kd *Discovery) Run(ch chan<- *config.TargetGroup, done <-chan struct{}) {
 	}
 }
 
+func (kd *Discovery) queryMasterPath(path string) (*http.Response, error) {
+	req, err := http.NewRequest("GET", path, nil)
+	if err != nil {
+		return nil, err
+	}
+	return kd.queryMasterReq(req)
+}
+
+func (kd *Discovery) queryMasterReq(req *http.Request) (*http.Response, error) {
+	// Lock in case we need to rotate masters to request.
+	kd.mastersMu.Lock()
+	defer kd.mastersMu.Unlock()
+	for i := 0; i < len(kd.Conf.Masters); i++ {
+		cloneReq := *req
+		cloneReq.URL.Host = kd.Conf.Masters[0].Host
+		cloneReq.URL.Scheme = kd.Conf.Masters[0].Scheme
+		res, err := kd.client.Do(&cloneReq)
+		if err == nil {
+			return res, nil
+		}
+		kd.rotateMasters()
+	}
+	return nil, fmt.Errorf("Unable to query any masters")
+}
+
+func (kd *Discovery) rotateMasters() {
+	if len(kd.Conf.Masters) > 1 {
+		kd.Conf.Masters = append(kd.Conf.Masters[1:], kd.Conf.Masters[0])
```

Comment:
Since `kd.Conf` is a `*KubernetesSDConfig` (a pointer to the originally read config), I think this will also backpropagate and rotate masters in the printed config on the status page, no? If so, that shouldn't happen. Maybe do the mutations on a copy of the masters list?

@fabxc we generally just rely on config-ed components to not modify their passed-in config, right?